File: python-service/src/controllers/sensor_controller.py
from typing import Optional, Dict, Any
from fastapi import HTTPException, Query, Request
from services.sensor_service import SensorService
from services.logger import Logger


class SensorController:
    def __init__(self, service: SensorService, logger: Logger):
        self.service = service
        self.logger = logger

    async def get_all_sensors(
        self,
        request: Request,
        type: Optional[str] = Query(None),
        limit: Optional[int] = Query(None),
        offset: int = Query(0)
    ) -> Dict[str, Any]:
        try:
            correlation_id = getattr(request.state, 'correlation_id', None)
            self.logger.set_correlation_id(correlation_id or '')

            filters = {'type': type, 'limit': limit, 'offset': offset}
            sensors = self.service.get_all_sensors(filters)
            
            return {'data': sensors, 'count': len(sensors)}
        except Exception as e:
            self.logger.error('Error fetching sensors', {'error': str(e)})
            raise HTTPException(status_code=500, detail='Internal server error')

    async def get_sensor_by_id(
        self,
        request: Request,
        sensor_id: str
    ) -> Dict[str, Any]:
        try:
            correlation_id = getattr(request.state, 'correlation_id', None)
            self.logger.set_correlation_id(correlation_id or '')

            return self.service.get_sensor_by_id(sensor_id)
        except ValueError as e:
            if 'not found' in str(e):
                raise HTTPException(status_code=404, detail='Sensor not found')
            raise HTTPException(status_code=400, detail=str(e))
        except Exception as e:
            self.logger.error('Error fetching sensor', {'error': str(e)})
            raise HTTPException(status_code=500, detail='Internal server error')

    async def create_sensor(
        self,
        request: Request,
        sensor_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        try:
            correlation_id = getattr(request.state, 'correlation_id', None)
            self.logger.set_correlation_id(correlation_id or '')

            created = self.service.create_sensor(sensor_data)
            return created
        except ValueError as e:
            if 'already exists' in str(e):
                raise HTTPException(status_code=409, detail=str(e))
            if 'out of valid range' in str(e) or 'must be' in str(e):
                raise HTTPException(status_code=400, detail=str(e))
            raise HTTPException(status_code=400, detail=str(e))
        except Exception as e:
            self.logger.error('Error creating sensor', {'error': str(e)})
            raise HTTPException(status_code=500, detail='Internal server error')

File: A1/node-service/src/controllers/sensorController.js
class SensorController {
  constructor(service, logger) {
    this.service = service;
    this.logger = logger;
  }

  async getAllSensors(req, res) {
    try {
      const correlationId = req.correlationId || req.headers['x-correlation-id'];
      this.logger.setCorrelationId(correlationId);

      // Extract query parameters (HTTP concern)
      const filters = {
        type: req.query.type,
        limit: req.query.limit ? parseInt(req.query.limit, 10) : undefined,
        offset: req.query.offset ? parseInt(req.query.offset, 10) : undefined
      };

      // Call service (business logic)
      const sensors = this.service.getAllSensors(filters);

      // Format HTTP response
      res.status(200).json({
        data: sensors,
        count: sensors.length
      });
    } catch (error) {
      this.logger.error('Error fetching sensors', { error: error.message });
      res.status(500).json({ error: 'Internal server error' });
    }
  }

  async getSensorById(req, res) {
    try {
      const correlationId = req.correlationId || req.headers['x-correlation-id'];
      this.logger.setCorrelationId(correlationId);

      const { id } = req.params; // Extract from URL path

      const sensor = this.service.getSensorById(id);

      res.status(200).json(sensor);
    } catch (error) {
      if (error.message === 'Sensor not found') {
        res.status(404).json({ error: 'Sensor not found' });
      } else {
        this.logger.error('Error fetching sensor', { error: error.message });
        res.status(500).json({ error: 'Internal server error' });
      }
    }
  }

  async createSensor(req, res) {
    try {
      const correlationId = req.correlationId || req.headers['x-correlation-id'];
      this.logger.setCorrelationId(correlationId);

      const sensorData = req.body; // Parse JSON body

      const created = this.service.createSensor(sensorData);

      res.status(201).json(created);
    } catch (error) {
      if (error.message === 'Sensor already exists') {
        res.status(409).json({ error: 'Sensor already exists' });
      } else if (error.message.includes('out of valid range') || error.message.includes('must be')) {
        res.status(400).json({ error: error.message });
      } else {
        this.logger.error('Error creating sensor', { error: error.message });
        res.status(400).json({ error: error.message });
      }
    }
  }
}

module.exports = SensorController;
